File: src/domain/entities/Patient.ts
import type { RefinementCtx } from "zod";
import { z } from "$i18n/zod";

export const patientSchema = z.object({
  birthdate: z
    .string()
    .transform((dateString: string, ctx: RefinementCtx) => {
      if (dateString) {
        const date = new Date(dateString);
        if (!z.date().safeParse(date).success) {
          ctx.addIssue({
            code: z.ZodIssueCode.invalid_date,
          });
        }
      }
      return dateString;
    })
    .nullable(),
  // birthdate: z.string().date().nullable(), // TODO wait for Zod new version release
  createdAt: z.date().nullable(),
  contactInformation: z.object({
    email: z.string().email().nullable(),
    phone: z.string().nullable(),
    mobilePhone: z.string().nullable(),
  }),
  diabetic: z.boolean().default(false),
  longDurationDisease: z.boolean().default(false),
  nationalInsuranceNumber: z.string().nullable(),
  notes: z.string().nullable(),
  personalName: z.object({
    firstName: z.string().min(2),
    lastName: z.string().min(2),
  }),
  postalAddress: z.object({
    city: z.string().min(2).nullable(),
    country: z.string().min(2).nullable(),
    postalCode: z.string().min(2).nullable(),
    street: z.string().min(2).nullable(),
  }),
  updatedAt: z.date().nullable(),
});

export type Patient = z.infer<typeof patientSchema>;

File: src-tauri/src/tests/domain/entity/patient.rs
use celes::Country;
use chrono::{NaiveDate, Utc};
use uuid::Uuid;

use crate::domain::{
    entity::patient::Patient,
    value_object::{
        contact_information::ContactInformation, personal_name::PersonalName,
        postal_address::PostalAddress,
    },
};

#[cfg(test)]

pub fn valid_patient_test_helper() -> Patient {
    Patient {
        birthdate: Some(NaiveDate::parse_from_str("1978-01-09", "%Y-%m-%d").unwrap()),
        created_at: Utc::now(),
        contact_information: ContactInformation {
            email: Some("jeremy.jousse@gmail".into()),
            mobile_phone: Some("0650805080".into()),
            phone: Some("032005500550".into()),
        },
        diabetic: false,
        id: Uuid::new_v4(),
        long_duration_disease: false,
        national_insurance_number: Some("017859000123".into()),
        notes: "My best patient".into(),
        personal_name: PersonalName {
            first_name: "Jérémy".into(),
            last_name: "Jousse".into(),
            title: None,
        },
        profession: Some("Developer".into()),
        postal_address: PostalAddress {
            city: Some("Lille".into()),
            country: Some(Country::france()),
            postal_code: Some("59000".into()),
            street: Some("1 rue de Paris".into()),
        },
        updated_at: Utc::now(),
    }
}
